promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+			if keywordIndex < 0 {
+				continue
+			}
+
+			if items[keywordIndex-1].Typ != parser.COMMENT && keywordIndex-1 != aggregatorIndex {
+				formatItems = true
+			} else if items[keywordIndex-1].Typ == parser.COMMENT {
+				// Peek back until no comment is found.
+				// If the item immediately before the comment is not aggregator, that means
+				// the keyword is not (immediately) after aggregator item and hence formatting
+				// is required.
+				var j int
+				for j = keywordIndex; j > aggregatorIndex; j-- {
+					if items[j].Typ == parser.COMMENT {
+						continue
+					}
+					break
+				}
+				if j != aggregatorIndex {
+					formatItems = true
+				}
+			}
+			if formatItems {
+				var (
+					tempItems              []parser.Item
+					finishedKeywordWriting bool
+				)
+				// Get index of the closing paren.
+				for j := keywordIndex; j <= len(items); j++ {
+					if items[j].Typ == parser.RIGHT_PAREN {
+						closingParenIndex = j
+						break
+					}
+				}
+				if closingParenIndex == -1 {
+					panic("invalid paren index: closing paren not found")
+				}
+				// Re-order lexical items. TODO: consider using slicing of lists.
+				for itemp := 0; itemp < len(items); itemp++ {
+					if itemp <= aggregatorIndex || itemp > closingParenIndex {
+						tempItems = append(tempItems, items[itemp])
+					} else if !finishedKeywordWriting {
+						// Immediately after writing the aggregator, write the keyword and
+						// the following expression till the closing paren. This is done
+						// to be in-order with the expected result.
+						for j := keywordIndex; j <= closingParenIndex; j++ {
+							tempItems = append(tempItems, items[j])
+						}
+						// itemp has not been updated yet and its holding the left paren.
+						tempItems = append(tempItems, items[itemp])
+						finishedKeywordWriting = true
+					} else if !(itemp >= keywordIndex && itemp <= closingParenIndex) {
+						tempItems = append(tempItems, items[itemp])
+					}
+				}
+				items = tempItems
+			}
+
+		case parser.IDENTIFIER:
+			//	TODO: currently its assumed that STRING are always label values.
+			itr := advanceComments(items, i)
+			if i < 1 || item.Val != "__name__" || items[itr+2].Typ != parser.STRING {
+				continue
+			}
+			var (
+				leftBraceIndex  = -1
+				rightBraceIndex = -1
+				labelValItem    = items[itr+2]
+				metricName      = labelValItem.Val[1 : len(labelValItem.Val)-1] // Trim inverted commas.
+				tmp             []parser.Item
+				skipBraces      bool // For handling metric_name{} -> metric_name
+			)
+			// We aim to convert __name__ to an ideal metric_name if the value of that labels is atomic.
+			// Since a non-atomic metric_name will contain alphabets other than a-z and A-Z including _,
+			// anything that violates this ceases the formatting of that particular label item.
+			// If this is not done then the output from the prettier might be an un-parsable expression.
+			if !regexp.MustCompile("^[a-z_A-Z]+$").MatchString(metricName) {
```

Comment:
If you declare regular Expressions, that `MustCompile` on a global scope, you have more reliable test failures in case they don't compile.

Reply:
Our requirements are limited to the values possessed by the label `__name__`. Regular expressions as metric_name should as such not be converted to ideal metric_name. I hope I got your question.

Reply:
> I hope I got your question.

It was more of a statement than a question. What I meant is that define this regular expression outside a function.

Reply:
Addressed it.

Reply:
> Addressed it.

Um, nothing has changed here, yet.

Maybe only resolve comments after pushing the respective changes.